chalice/app.py

```diff
@@ -266,15 +292,42 @@ def __init__(self, body, headers=None, status_code=200):
         self.headers = headers
         self.status_code = status_code
 
-    def to_dict(self):
+    def to_dict(self, binary_types=None):
         body = self.body
-        if not isinstance(body, str):
+        if not isinstance(body, (str, bytes)):
             body = json.dumps(body, default=handle_decimals)
-        return {
+        response = {
             'headers': self.headers,
             'statusCode': self.status_code,
-            'body': body,
+            'body': body
         }
+        if binary_types is not None:
+            self._b64encode_body_if_needed(response, binary_types)
+        return response
+
+    def _b64encode_body_if_needed(self, response_dict, binary_types):
+        response_headers = CaseInsensitiveMapping(response_dict['headers'])
+        content_type = response_headers.get('content-type', '')
+        body = response_dict['body']
+
+        # Encode the response as binary if required.
+        # To keep json types working as expected they are encoded to a utf-8
+        # byte sequence if the content type is application/json before being
+        # converted to a base64 string.
+        if _matches_content_type(content_type, binary_types):
+            if _matches_content_type(content_type, ['application/json']):
```

Comment:
JSON is automatically handled for the user, so there is a special case when they make `application/json` a binary type. Since the contract for `application/json` is you can send any json serializable type as the `body` and it will work. But for a binary type you must send a `bytes()` type. So if you specify `application/json` as a binary type, we do the conversion first, which results in a string. Not base64 encodable in py3. Since JSON is utf-8 we can encode it using that to a bytes type and then base64 encode that.

That way you can still send a dict/list/jsonish type as a body and have a json binary type and it will still do the right thing.